src/lib/mongoBackend/mongoConnectionPool.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+*
+* Copyright 2015 Telefonica Investigacion y Desarrollo, S.A.U
+*
+* This file is part of Orion Context Broker.
+*
+* Orion Context Broker is free software: you can redistribute it and/or
+* modify it under the terms of the GNU Affero General Public License as
+* published by the Free Software Foundation, either version 3 of the
+* License, or (at your option) any later version.
+*
+* Orion Context Broker is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Ken Zangelin
+*/
+#include <time.h>
+#include <semaphore.h>
+
+#include "logMsg/logMsg.h"
+#include "common/clockFunctions.h"
+#include "mongoBackend/mongoConnectionPool.h"
+#include "mongoBackend/MongoGlobal.h"
+
+
+
+/* ****************************************************************************
+*
+* MongoConnection - 
+*/
+typedef struct MongoConnection
+{
+  DBClientBase*  connection;
+  bool           free;
+} MongoConnection;
+
+
+
+/* ****************************************************************************
+*
+* globals - 
+*/
+static MongoConnection* connectionPool     = NULL;
+static int              connectionPoolSize = 0;
+static sem_t            connectionPoolSem;
+static sem_t            connectionSem;
+static struct timespec  semWaitingTime     = { 0, 0 };
+static bool             semStatistics      = false;
+
+
+
+/* ****************************************************************************
+*
+* mongoConnectionPoolInit - 
+*/
+int mongoConnectionPoolInit
+(
+  const char* host,
+  const char* db,
+  const char* rplSet,
+  const char* username,
+  const char* passwd,
+  double      timeout,
+  int         writeConcern,
+  int         poolSize,
+  bool        semTimeStat
+)
+{
+  //
+  // Checking incoming parameters
+  //
+  if (poolSize == 0)
```

Comment:
It happened to me and EVERYTHING stopped working.
It was a bug of course, but I prefer to protect the code from bugs like that. It took me 50 LM_M and a few hours to understand the whole thing ...

NTC